mesonbuild/backend/ninjabackend.py

```diff
@@ -992,15 +992,14 @@ def generate_vala_compile(self, target, outfile):
             # Library name
             args += ['--library=' + target.name]
             # Outputted header
-            hname = os.path.join(self.get_target_dir(target), target.name + '.h')
+            hname = os.path.join(self.get_target_dir(target), target.vala_header)
             args += ['-H', hname]
             valac_outputs.append(hname)
             # Outputted vapi file
-            base_vapi = target.name + '.vapi'
-            vapiname = os.path.join(self.get_target_dir(target), base_vapi)
+            vapiname = os.path.join(self.get_target_dir(target), target.vala_vapi)
             # Force valac to write the vapi file in the target build dir.
             # Without this, it will write it inside c_out_dir
-            args += ['--vapi=../' + base_vapi]
+            args += ['--vapi=../' + target.vala_vapi]
```

Comment:
The two are not the same. `get_target_dir()` will return the subdir relative to the build root where the target is being built (the "target build dir"), which is not `..`.

For some reason, `valac` does not put the generated `.h` and the `.vapi` in the same directory. The `.h` file is placed in the directory as specified by the path, but the `.vapi` is placed alongside the generated C files in the 'target private directory' (see `c_out_dir` above) which is a subdir of the target build dir. The `..` is a hack to make it output the `.vapi` in the target build dir.

Try it out and see. You can make it `os.path.join('..', target.vala_vapi)` if you wish. :)